pkg/k8sutil/portforward.go

```diff
@@ -26,39 +27,90 @@ import (
 	"k8s.io/client-go/transport/spdy"
 )
 
-func IsPortAvailable(port int) bool {
+func IsPortAvailable(clientset kubernetes.Interface, port int) (bool, error) {
+	// kube-proxy stopped opening ports on the host for NodePort services in recent versions of kubernetes,
+	// so net.Listen won't be able to detect if the port is used by a NodePort service or not, and port forwarding might hang or redirect to the wrong component.
+	// https://github.com/kubernetes/kubernetes/pull/108496
+	// so we check if the port is used by a NodePort service.
+	services, err := clientset.CoreV1().Services("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// TODO: make this work with minimal RBAC
+		if !kuberneteserrors.IsForbidden(err) {
+			return false, errors.Wrap(err, "failed to list services")
+		}
+	}
+
+	for _, service := range services.Items {
+		if service.Spec.Type != corev1.ServiceTypeNodePort {
+			continue
+		}
+		for _, p := range service.Spec.Ports {
+			if p.NodePort > 0 && int(p.NodePort) == port {
+				return false, nil
+			}
+		}
+	}
+
 	// portforward explicitly listens on localhost
 	host := net.JoinHostPort("127.0.0.1", strconv.Itoa(port))
 	server, err := net.Listen("tcp4", host)
 	if err != nil {
-		return false
+		return false, nil
 	}
 
 	server.Close()
-	return true
+	return true, nil
+}
+
+func FindFreePort(clientset kubernetes.Interface) (int, error) {
+	port, err := freeport.GetFreePort()
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to get free port")
+	}
+
+	isPortAvailable, err := IsPortAvailable(clientset, port)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check if port is available")
+	}
+
+	if !isPortAvailable {
+		return FindFreePort(clientset)
```

Comment:
Wanna put a limit on this recursion?  If `isPortAvailable` stays false, this will spin forever.

Reply:
yeah, i knew you were gonna say that... okay :)